chainercv/experimental/links/model/fcis/fcis.py

```diff
@@ -0,0 +1,308 @@
+
+        If the attributes need to be changed to something
+        other than the values provided in the presets, please modify
+        them by directly accessing the public attributes.
+
+        Args:
+            preset ({'visualize', 'evaluate'): A string to determine the
+                preset to use.
+
+        """
+        if preset == 'visualize':
+            self.nms_thresh = 0.3
+            self.score_thresh = 0.7
+            self.mask_merge_thresh = 0.5
+            self.binary_thresh = 0.4
+            self.limit = 100
+            self.min_drop_size = 16
+        elif preset == 'evaluate':
+            self.nms_thresh = 0.3
+            self.score_thresh = 1e-3
+            self.mask_merge_thresh = 0.5
+            self.binary_thresh = 0.4
+            self.limit = 100
+            self.min_drop_size = 16
+        else:
+            raise ValueError('preset must be visualize or evaluate')
+
+    def predict(self, imgs):
+        """Segment object instances from images.
+
+        This method predicts instance-aware object region for each image.
+
+        Args:
+            imgs (iterable of numpy.ndarray): Arrays holding images of shape
+                :math:`(B, C, H, W)`.  All images are in CHW and RGB format
+                and the range of their value is :math:`[0, 255]`.
+
+        Returns:
+           tuple of lists:
+           This method returns a tuple of three lists,
+           :obj:`(masks, labels, scores)`.
+
+           * **masks**: A list of float arrays of shape :math:`(R, H, W)`, \
+               where :math:`R` is the number of masks in a image. \
+               Each pixel holds value if it is inside the object inside or not.
+           * **labels** : A list of integer arrays of shape :math:`(R,)`. \
+               Each value indicates the class of the masks. \
+               Values are in range :math:`[0, L - 1]`, where :math:`L` is the \
+               number of the foreground classes.
+           * **scores** : A list of float arrays of shape :math:`(R,)`. \
+               Each value indicates how confident the prediction is.
+
+        """
+
+        prepared_imgs = []
+        sizes = []
+        for img in imgs:
+            size = img.shape[1:]
+            img = self.prepare(img.astype(np.float32))
+            prepared_imgs.append(img)
+            sizes.append(size)
+
+        masks = []
+        labels = []
+        scores = []
+
+        for img, size in zip(prepared_imgs, sizes):
+            with chainer.using_config('train', False), \
+                    chainer.function.no_backprop_mode():
+                # inference
+                img_var = chainer.Variable(self.xp.array(img[None]))
+                scale = img_var.shape[3] / size[1]
+                roi_seg_scores, _, roi_scores, rois, _ = \
+                    self.__call__(img_var, scale)
+
+            roi_seg_score = roi_seg_scores.array
+            roi_score = roi_scores.array
+            bbox = rois / scale
```

Comment:
How about not introducing `bbox`.
I mean use `roi = rois / scale`.

The rationale to prefer `roi` is explained in the second suggestion of https://github.com/chainer/chainercv/pull/568#discussion_r183230729.

Reply:
By definition, `rois ` and `bboxes` are different (`bboxes` is a list of `bbox`).
But, as you said, in the case when `batchsize=1`, `rois = bboxes`.

Having said that, I think that continuing to use `rois` in `predict` is better because it explicitly indicates that `roi_mask_prob` is defined relative to `rois`.
This is the same reason why we use `rois` elsewhere, such as `__call__`.

Reply:
Good?

Reply:
I don't see the difference of `bboxes` and `rois`.
In my understanding, `bboxes` and `rois` have same shape `(R, 4)`.
In FCIS, we don't use `roi_ag_locs` (because it is already applied in `__call__` with `iter2=True`) so that `roi` is same as `bbox`.
For the return output, the `predict` should return `bbox` as output not `roi`, so that I rename here.

Reply:
After looking back at implementations of detection, we have no strict rule on when we can remove `roi_`. I understand that it is convenient to use shortend notation (e.g., `roi_prob -> prob`).

My main concern was that subset of variable has `roi_`, because it looks inconsistent (I would look for `roi` that corresponds to `roi_mask_score`).

https://github.com/knorth55/chainercv/blob/da631a105794f50e54e9cf991c788d74ed2b1463/chainercv/experimental/links/model/fcis/fcis.py#L278

To solve this problem, I want the code to transit from stop appending `roi_*` for all variables at same time.
How about using `roi`, `roi_seg_prob` and `roi_prob` until L278, and using `bbox`, `roimask_prob`, `label` and `prob` after L278 (the return vales of `mask_voting`)?
`roimask` is a name for data type of `(R, RH, RW)`.

Reply:
> No. bboxes is a list of (R, 4), whereas rois is (R', 4).

You mean `bbox` is after non maximum suppression and other processing?
In that case, this `rois` are already `bboxes`.

Reply:
`roi_mask` are complete different from other `roi_*`.
What you want is removing other `roi_*` variables. Maybe i understand.
In that case, the rule below sounds good to me
```
roi_mask -> roi_mask (no change)
rois -> bboxes
roi_seg_scores -> seg_scores
roi_scores -> scores
```

Reply:
What I was saying is this.

```
roi_mask -> roimask
rois -> bboxes
roi_seg_scores -> seg_scores
roi_scores -> scores
```

Reply:
> You mean bbox is after non maximum suppression and other processing?
In that case, this rois are already bboxes.

No no. The types are different.

Reply:
> No no. The types are different.

Types are different? Can  you tell me more detailed information?

>roi_mask -> roimask

And I prefer `roi_mask` to `roimask` because it is easy to read.